### PR TITLE
feat: enable full foreign collect flow via extrinsic

### PR DESCRIPTION
**Pull request by wischli**
# Description

## Collect

* Adds `InvestmentPaymentCurrency` storage for the `foreign payment currency` upon receiving the first IncreaseInvestOrder message (similar to redemptions)
  * `InvestmentPaymentCurrency` will be cleared if the `InvestState` is cleared, i.e. if neither an unprocessed swap nor investment is happening in case of reaching `NoState` or `SwapIntoForeignDone`
  * `InvestmentPaymentCurrency` will be overwritten upon receiving `IncreaseInvestOrder` or `CollectInvestOrder` to be less restrictive to investors
  * We will read from that storage after the investment was collected through `pallet_investments::collect_invest_for` such that the dispatch of `ExecutedCollectInvest` happens automatically
* This enables us to not have to add a `collect_foreign_investment_for` extrinsic.
* For collecting redemptions, this flow is already implemented. However, there is an async gap in between collecting the redemption (into `pool `currency) and fully fulfilling the swap order (to `foreign` currency).
  * When that swap is fulfilled, the dispatch of `ExecutedCollectRedeem` happens automatically
* Of course, triggering collections through LP messages from EVM domains is also supported. The flows would be the same as for `pallet_investments::collect_{invest, redeem}_for`.

## Misc

* Renames `allow_pool_currency` extrinsic to `allow_investment_currency`
* Slightly reduces `foreign_investments` integration test boilerplate

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: pallets/foreign-investments/src/impls/mod.rs [hunk 1]
```diff
@@ -56,6 +54,11 @@ impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
 			!T::Investment::investment_requires_collect(who, investment_id),
 			Error::<T>::InvestError(InvestError::CollectRequired)
 		);
+
+		// NOTE: For the MVP, we simply overwrite any existing payment currency. In a
+		// later version, we might want to store a `BoundedVec` instead.
```
Comment (NunoAlexandre) [inline pallets/foreign-investments/src/impls/mod.rs:59]: I almost always find it better to use a double map instead with `()` as the value type since that makes storage reads/writes a lot cheaper and keeps the code simple as it doesn't need the Max for the bounded vec. If this is something we may need in the future, it's low effort to implement it already and save the migration work later on.

File: pallets/foreign-investments/src/impls/mod.rs [hunk 2]
```diff
@@ -325,6 +328,10 @@ impl<T: Config> Pallet<T> {
 		match state {
 			InvestState::NoState => {
 				InvestmentState::<T>::remove(who, investment_id);
+				#[cfg(feature = "std")] {
+					println!("Clearing InvestState");
+				}
```
Comment (NunoAlexandre) [inline pallets/foreign-investments/src/impls/mod.rs:333]: Of curiosity, is it possible to use the `log` instead so that we can have these logs both on std and wasm?

File: pallets/foreign-investments/src/lib.rs [hunk 3]
```diff
@@ -412,17 +437,12 @@ pub mod pallet {
...
-	// 		fn from(error: InvestError) -> Self {
-	// 			Error::<T>::InvestError(error)
-	// 		}
-	// 	}
-
-	// 	impl<T> From<RedeemError> for Error<T> {
-	// 		fn from(error: RedeemError) -> Self {
-	// 			Error::<T>::RedeemError(error)
-	// 		}
-	// 	}
```
Comment (NunoAlexandre) [inline pallets/foreign-investments/src/lib.rs:427]: 🧹 ✨

File: pallets/liquidity-pools/src/hooks.rs [hunk 4]
```diff
@@ -101,7 +104,57 @@ where
...
+		};
+
+		T::OutboundQueue::submit(T::TreasuryAccount::get(), domain_address.domain(), message)?;
+
+		Ok(())
+	}
+}
+
+/// The hook struct which acts upon a finalized investment collection.
+
```
Comment (NunoAlexandre) [inline pallets/liquidity-pools/src/hooks.rs:117]:
```suggestion
```

File: runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs [hunk 5]
```diff
@@ -634,7 +647,8 @@ mod same_currencies {
 				1
 			);
 
-			// User collects through foreign investments
+			// Should not mutate any state if user collects through foreign investments
+			// after collecting through Investments
```
Comment (NunoAlexandre) [inline runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs:651]: these comments are really helpful 🙏

File: runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs [hunk 6]
```diff
@@ -1450,10 +1477,96 @@ mod mismatching_currencies {
...
+			let pool_id = DEFAULT_POOL_ID;
+			let investor: AccountId =
+				AccountConverter::<DevelopmentRuntime, LocationToAccountId>::convert((
+					DOMAIN_MOONBEAM,
+					BOB,
+				));
+			let pool_currency: CurrencyId = AUSD_CURRENCY_ID;
+			let foreign_currency: CurrencyId = USDT_CURRENCY_ID;
+			let pool_currency_decimals = currency_decimals::AUSD;
+			let invest_amount_pool_denominated: u128 = 6_000_000_000_000_000;
```
Comment (NunoAlexandre) [inline runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs:1503]: can we use the `dollar(<decimals>)` function here? So this would become `6 * dollar(18)` (where 18 is the AUSD decimals. Even better if we have a constant for that). Just to facilitate reading these amounts and to avoid a situation where for example we may be moving amounts that are too high or too low and not quite testing things properly.

File: runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs [hunk 7]
```diff
@@ -1450,10 +1477,96 @@ mod mismatching_currencies {
...
+			let sending_domain_locator = Domain::convert(DEFAULT_DOMAIN_ADDRESS_MOONBEAM.domain());
+			create_currency_pool(pool_id, pool_currency, pool_currency_decimals.into());
+			let invest_amount_foreign_denominated: u128 =
+				enable_usdt_trading(pool_currency, invest_amount_pool_denominated, true, || {});
+			do_initial_increase_investment(
```
Comment (NunoAlexandre) [inline runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs:1508]: Nit but personally as a reader I like to have a line break between all the `lets` and an actual call 🙏 

```suggestion
			
			do_initial_increase_investment(
```

File: runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs [hunk 8]
```diff
@@ -1450,10 +1477,96 @@ mod mismatching_currencies {
...
+				),
+				Some(foreign_currency)
+			);
+			assert!(Tokens::balance(default_investment_id().into(), &investor).is_zero());
+			assert_eq!(
+				Tokens::balance(default_investment_id().into(), &sending_domain_locator),
+				invest_amount_pool_denominated * 2
+			);
+
+			// Should not be cleared as invest state is swapping into pool currency
```
Comment (NunoAlexandre) [inline runtime/integration-tests/src/liquidity_pools/pallet/development/tests/liquidity_pools/foreign_investments.rs:1559]: Again, super nice to have these checks and these comments 👌

**Review by NunoAlexandre — DISMISSED** [review]
Very nice work, like always 💯 

A couple of small comments but no blockers for me.

File: pallets/foreign-investments/src/impls/mod.rs (hunk 2 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:333]: Sorry, this is an unwanted remnant from debugging integration tests. I will remove it!

File: pallets/foreign-investments/src/impls/mod.rs (hunk 1 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:59]: I think both options have separate use cases. When using a `DoubleStorageMap`, you don't want to limit the key-value pairs which you want when using a `StorageMap` with `BoundedVec` values. 

Nevertheless, I agree a double map might be more suitable here. Given that we currently limit `(investor, investment_id)` pairs to have at most one swap at the same time, adding true multicurrency support (both in EVM as well as Centrifuge Chain) will be a bigger feature. It was discussed this is something we need mid- to long-term (rather 2024) but not for this MVP.

Comment (NunoAlexandre) [reply pallets/foreign-investments/src/impls/mod.rs]: > [It was discussed ](https://centrifuge.hackmd.io/tKGS5CwqSQeeI3bU1dKUlw#20230815)this is something we need mid- to long-term (rather 2024) but not for this MVP.


Then I say we do this now, it takes very little effort compared to the significant effort and error-proneness that writing a migration for this in the future will entail.

Comment (NunoAlexandre) [reply pallets/foreign-investments/src/impls/mod.rs]: I see it's done already, thanks for doing that :)

**Review by NunoAlexandre — APPROVED** [review]
👌  nice

File: pallets/foreign-investments/src/impls/mod.rs (hunk 1 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs]: I just updated the comment but did not alter any storage. Migrating to double maps will touch multiple storages and complicates swap order handling much more. It definitely should not be part of this PR.

Comment (NunoAlexandre) [reply pallets/foreign-investments/src/impls/mod.rs]: Oh, I looked at the `InvestmentPaymentCurrency` which is a double map, my bad. I don't agree with the decision here but it's not a blocker for me.

File: libs/types/src/investments.rs [hunk 9]
```diff
@@ -185,7 +185,7 @@ pub struct Swap<
 	pub currency_in: Currency,
 	/// The outgoing currency, i.e. the one which should be replaced.
 	pub currency_out: Currency,
-	/// The amount of outgoing currency which shall be exchanged.
+	/// The amount of incoming currency which shall be bought.
```
Comment (wischli) [inline libs/types/src/investments.rs:188]: See [this comment](https://github.com/centrifuge/centrifuge-chain/pull/1509#discussion_r1331346309) for explanation.

**Comment by mustermeiszer** [comment]
Checking something from the description:

> InvestmentPaymentCurrency will be overwritten upon receiving IncreaseInvestOrder or CollectInvestOrder to be less restrictive to investors

Wouldn't that also mean, that a user could overwrite this storage with another currency mid-flight and for example, change the actual currency that is signaled via `ExecutedCollectInvest`?

File: pallets/foreign-investments/src/impls/mod.rs [hunk 10]
```diff
@@ -187,19 +190,19 @@ impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
 		who: &T::AccountId,
 		investment_id: T::InvestmentId,
 		foreign_payout_currency: T::CurrencyId,
-		pool_currency: T::CurrencyId,
-	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+	) -> DispatchResult {
+		// Simply overwrite any existing payment currency to enable the investor to have
+		// full control over the payment currency
+		InvestmentPaymentCurrency::<T>::insert(who, investment_id, foreign_payout_currency);
```
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls/mod.rs:196]: Isn't that criticial if something is happening currently? E.g.  swapping due to an `DecreaseInvestOrder`?

**Comment by wischli** [comment]
> Checking something from the description:
> 
> > InvestmentPaymentCurrency will be overwritten upon receiving IncreaseInvestOrder or CollectInvestOrder to be less restrictive to investors
> 
> Wouldn't that also mean, that a user could overwrite this storage with another currency mid-flight and for example, change the actual currency that is signaled via `ExecutedCollectInvest`?

You are right, shoot. Then, I don't see another way of limiting an `(investor, investment_id)` pair to the same foreign currency as long as the foreign investment state is not cleared which happens if neither an active (invest) swap nor an unprocessed amount exist. WDYT?

File: pallets/foreign-investments/src/impls/mod.rs (hunk 10 above)
Comment (mustermeiszer) [reply pallets/foreign-investments/src/impls/mod.rs:196]: Or a `IncreaseInvestOrder` with `payment_currency_a` followed by an `CollectInvest` with `payment_currency_b`

File: pallets/foreign-investments/src/impls/mod.rs [hunk 11]
```diff
@@ -1171,9 +1151,56 @@ impl<T: Config> Pallet<T> {
...
+	/// currency via the `CurrencyConverter` trait.
+	pub(crate) fn notify_executed_collect_invest(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+	) -> DispatchResult {
+		let foreign_payout_currency = InvestmentPaymentCurrency::<T>::get(who, investment_id)
+			.ok_or(Error::<T>::InvestmentPaymentCurrencyNotFound)?;
+		let pool_currency = T::PoolInspect::currency_for(investment_id.of_pool())
+			.ok_or(Error::<T>::PoolNotFound)?;
+		let collected = CollectedInvestment::<T>::take(who, investment_id);
```
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls/mod.rs:1169]: Due to being a `ValueQuery` this will always hit. Are we sure the solidity side wants to be informed if nothing was acutally transferred? 

Due to the need of bridging this seems like a big overhead for no information gain.

File: pallets/foreign-investments/src/impls/mod.rs [hunk 12]
```diff
@@ -1171,9 +1151,56 @@ impl<T: Config> Pallet<T> {
...
+	/// `ExecutedCollectInvest`.
+	///
+	/// NOTE: Converts the collected pool currency payment amount to foreign
+	/// currency via the `CurrencyConverter` trait.
+	pub(crate) fn notify_executed_collect_invest(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+	) -> DispatchResult {
+		let foreign_payout_currency = InvestmentPaymentCurrency::<T>::get(who, investment_id)
+			.ok_or(Error::<T>::InvestmentPaymentCurrencyNotFound)?;
```
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls/mod.rs:1166]: FYI: All these errors can be directly merged into the storage type now.

**Comment by wischli** [comment]
> You are right, shoot. Then, I don't see another way of limiting an `(investor, investment_id)` pair to the same foreign currency as long as the foreign investment state is not cleared which happens if neither an active (invest) swap nor an unprocessed amount exist. WDYT?

EDIT: Thinking about this more I actually think that's not really possible since collecting a redemption is synchronous. So it could only occur if within the same block an investor changes the `foreign_currency` by either increasing their investment or calling `collect` and this transaction is included before the `investment::collect_redeem_for` one. I don't see any issues with that as this signal's the investor's intend.

File: pallets/liquidity-pools/src/inbound.rs [hunk 13]
```diff
@@ -305,44 +303,13 @@ where
 		tranche_id: T::TrancheId,
 		investor: T::AccountId,
 		currency_index: GeneralCurrencyIndexOf<T>,
-		destination: DomainAddress,
 	) -> DispatchResult {
 		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
-		let currency_index_u128 = currency_index.index;
 		let payout_currency = Self::try_get_payout_currency(invest_id.clone(), currency_index)?;
```
Comment (mustermeiszer) [inline pallets/liquidity-pools/src/inbound.rs:308]: **CR**

Looking at the code of `try_get_payout_currency` I think this is not working. 

The check that is being used `T::ForeignInvestment::accepted_payment_currency(invest_id,currency)` will check the wrong trading pair. I.e. the wrong trading pair direction. As trading pairs are not ensured to work in both ways, this might return ok also it will fail when trying to swap.

**Review by mustermeiszer — CHANGES_REQUESTED** [review]
1. Unsure whether overwriting the invest currency is not potentially dangerous.
2. Quite sure that the check for `payout_currency` is currently a bug

**Comment by mustermeiszer** [comment]
> Thinking about this more I actually think that's not really an issue since collecting an investment is synchronous. So it could only occur if within the same block

But I think given the current code complexity, this is quite an assumption for the future. I mean it being syncrhonous all the time. 

And IIRC we wanted that for the Solidity side and if IIRC this does not handle it well

Example:
* a user invests from multiple `LP`s  (`Lp(currency_a)`, `LP(currency_b)`)
* user triggers `IncreaseInvestOrder(currency_a)` &rarr; `msg(3)`
* user triggers `CollectInvest(currency_a)` &rarr; `msg(2)`
* user triggers `IncreaseInvestOrder(currency_b)` &rarr; `msg(3)`

Message execution on our side due to bridge or mem-pool is: 
* `msg(1)`
* `msg(3)`
* `msg(2)`

this would mean the wrong `LP` on the Solidity side is informed about the collect and the right one never sees it. But maybe I am also not getting the Solidity side correctly. cc @hieronx and @ilinzweilin

File: pallets/foreign-investments/src/impls/mod.rs (hunk 11 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:1169]: I am confused, in #1473 [you requested](https://github.com/centrifuge/centrifuge-chain/pull/1473#discussion_r1323224830) to allow collecting zero amounts, so I removed the barrier. IIUC, now you contradict your previous request?

File: pallets/liquidity-pools/src/inbound.rs (hunk 13 above)
Comment (wischli) [reply pallets/liquidity-pools/src/inbound.rs:308]: Great catch! Will fix `fn try_get_payout_currency`.

**Comment by wischli** [comment]
> But I think given the current code complexity, this is quite an assumption for the future. I mean it being syncrhonous all the time.

When would it not be synchronous? Collecting an investment does not require a swap. It just performs `T::Investment::collect_invest_for` and puts `CollectInvest` into the `OutboundQueue` based on the information stored during the `collect_invest_for` and the `InvestmentPaymentCurrency`.

Regarding the example, your message id declaration is off. I suppose you wanted the first message to have index 1.

> And IIRC we wanted that for the Solidity side and if IIRC this does not handle it well
> 
> Example:
> 
> * a user invests from multiple `LP`s  (`Lp(currency_a)`, `LP(currency_b)`)
> * user triggers `IncreaseInvestOrder(currency_a)` → ~~`msg(3)`~~ `msg(1)`
> * user triggers `CollectInvest(currency_a)` → `msg(2)`
> * user triggers `IncreaseInvestOrder(currency_b)` → `msg(3)` 

If the user invests from multiple Liquidity Pools, each investment has their separate `investment_id` such that these are considered separate investments `(investor, LP(currency_a))` and `(investor, LP(currency_b))`.

If we had that message flow for the same `(investor, investment_id)` (i.e. same tranche id of the same LP), then you are right. I blanked out on the fact that we only need to do this so we can map the message to the correct contract address. I only thought about the response we give to the investor. And if that used `currency_b` instead of `currency_a`, it would just translate the payment amount into `currency_b` instead of `currency_a`.

TLDR: We need to restrict an investor to the same payment currency until the foreign investment state is cleared for now on Centrifuge chain IF we want to enable the full `collect_foreign_investment` flow by just calling `collect_invest_for` in `pallet-investment`. If that restriction is plausible, then I will make the appropriate adjustments to this PR. Please let me know what you think @mustermeiszer @hieronx

File: pallets/foreign-investments/src/impls/mod.rs (hunk 11 above)
Comment (mustermeiszer) [reply pallets/foreign-investments/src/impls/mod.rs:1169]: No, I am not. I am just saying that if we trigger a return message, we should be sure Solidity actually needs it. If not we should collect and not trigger a return message. But lets just keep it as is. Its just optimisation.

File: pallets/liquidity-pools/src/inbound.rs (hunk 13 above)
Comment (mustermeiszer) [reply pallets/liquidity-pools/src/inbound.rs:308]: I can not, but it might still happen by accident as no logic in the order book ensures bidirectionality/

**Comment by mustermeiszer** [comment]
@hieronx need to answer if this is sufficient. I would for now maybe check on-chain that if an `InvestmentPaymentCurrency` is set, all new updates must be of the same currency, wdyt? Then the chain is save from wrongly generated message on the Solidity or frontend side.

File: pallets/foreign-investments/src/impls/mod.rs (hunk 11 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:1169]: I have been favouring blocking a return message if the amount is zero. I will implement this.

File: pallets/liquidity-pools/src/inbound.rs (hunk 13 above)
Comment (wischli) [reply pallets/liquidity-pools/src/inbound.rs:308]: On that note: If bidirectionality is not given, an investor cannot decrease their foreign investment if it creates an orderbook swap from `pool` to `foreign` because then `TradingPair::<T>::get(pool_currency, foreign_currency)`  does not exist. Moreover, we block redemptions for that `foreign` currency as well.

AFAICT, we should check in `fn allow_investment_currency` both if the currency is accepted as payment as well as payout, not only the former. WDYT?

File: pallets/foreign-investments/src/impls/mod.rs (hunk 11 above)
Comment (hieronx) [reply pallets/foreign-investments/src/impls/mod.rs:1169]: I think there is a possible scenario in which someone submitted a redemption order, and then the price ended up going to 0, and the redemption is fully executed but with 0 currency payout. In that case, the value of sending `ExecutedCollectRedeem` to the other domains is that they will update the local `remaining_redeem_order` state to 0.

**Comment by hieronx** [comment]
> TLDR: We need to restrict an investor to the same payment currency until the foreign investment state is cleared for now on Centrifuge chain IF we want to enable the full collect_foreign_investment flow by just calling collect_invest_for in pallet-investment.
> I would for now maybe check on-chain that if an InvestmentPaymentCurrency is set, all new updates must be of the same currency, wdyt? 

Seems totally fine to me 👍

File: pallets/liquidity-pools/src/inbound.rs (hunk 13 above)
Comment (wischli) [reply pallets/liquidity-pools/src/inbound.rs:308]: > AFAICT, we should check in fn allow_investment_currency both if the currency is accepted as payment as well as payout, not only the former. WDYT?

cc @hieronx

Comment (hieronx) [reply pallets/liquidity-pools/src/inbound.rs:308]: Yes 100%. I think in practice governance should ensure that always both directions of an LP token trading pair are added.

File: pallets/foreign-investments/src/impls/mod.rs (hunk 11 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:1195]: > I think there is a possible scenario in which someone submitted a redemption order, and then the price ended up going to 0, and the redemption is fully executed but with 0 currency payout. In that case, the value of sending `ExecutedCollectRedeem` to the other domains is that they will update the local `remaining_redeem_order` state to 0.

IIUC, the payment amount should then not be zero. I added a short circuit if the both the accumulated payment amount as well as collected amount are zero: [0b6924d](https://github.com/centrifuge/centrifuge-chain/pull/1556/commits/0b6924df3d14d6c18a15f760746259baca6b47c2)

File: pallets/foreign-investments/src/impls/mod.rs (hunk 10 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs]: [2768b4b](https://github.com/centrifuge/centrifuge-chain/pull/1556/commits/2768b4bdd81be4f3d025881252b75f6f0b538d08)

File: pallets/foreign-investments/src/impls/mod.rs [hunk 14]
```diff
@@ -186,20 +213,22 @@ impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
 	fn collect_foreign_investment(
 		who: &T::AccountId,
 		investment_id: T::InvestmentId,
-		foreign_payout_currency: T::CurrencyId,
-		pool_currency: T::CurrencyId,
-	) -> Result<ExecutedForeignCollectInvest<T::Balance>, DispatchError> {
+		foreign_payment_currency: T::CurrencyId,
+	) -> DispatchResult {
+		let payment_currency = InvestmentPaymentCurrency::<T>::get(who, investment_id)?;
```
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls/mod.rs:218]: Just wandering regarding the states of this. Wouldn't it be possible for the initial `IncreaseInvestOrder` to be fully "done" before a `CollectInvest` is actually triggered? 

Meaning, the `InvestmentPaymentCurrency` state would be `None`.

File: pallets/foreign-investments/src/impls/mod.rs [hunk 15]
```diff
@@ -209,14 +238,10 @@ impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
 		foreign_payout_currency: T::CurrencyId,
 		pool_currency: T::CurrencyId,
 	) -> Result<(), DispatchError> {
+		let payout_currency = RedemptionPayoutCurrency::<T>::get(who, investment_id)?;
```
Comment (mustermeiszer) [inline pallets/foreign-investments/src/impls/mod.rs:241]: Not sure if the same thinking applies to redemptions.

**Review by mustermeiszer — COMMENTED** [review]
All looks good. I got a minor logical worry regarding the need to have a valid `InvestmentPaymentCurrency` when collecting. If this is unjustified it is good to go.

File: pallets/foreign-investments/src/impls/mod.rs (hunk 15 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:241]: A `RedemptionState` is only cleared if 
1. either the entire unprocessed amount is decreased without an active/done swap (`Redeeming` --> `NoState`) or
2. the entire swap is fulfilled without unprocessed redeeming amount (`SwapIntoForeignDone` --> `NoState`)

The first one is irrelevant here because it requires the `RedemptionState` to be re-initialized by increasing the redemption which sets the payout currency.

The second one triggers the redemption collection transition before/when clearing:

https://github.com/centrifuge/centrifuge-chain/blob/0b6924df3d14d6c18a15f760746259baca6b47c2/pallets/foreign-investments/src/impls/mod.rs#L503-L509

https://github.com/centrifuge/centrifuge-chain/blob/0b6924df3d14d6c18a15f760746259baca6b47c2/pallets/foreign-investments/src/impls/mod.rs#L706-L714

File: pallets/foreign-investments/src/impls/mod.rs (hunk 14 above)
Comment (wischli) [reply pallets/foreign-investments/src/impls/mod.rs:218]: An `InvestmentState` is only cleared if
1. either the entire unprocessed amount was decreased and that swap is fulfilled (`SwapIntoForeignDone` --> `NoState`)
2. The state only holds an unprocessed investment amount which was fully processed (`InvestmentOngoing` --> `NoState`)

The first one is irrelevant here because it requires the `InvestmentState` to be re-initialized by increasing the investment which sets the payment currency.

The second one is only known if the investment is collected. Therefore, we trigger the foreign collect flow before clearing the `InvestmentState` in `apply_invest_state_transition`:

https://github.com/centrifuge/centrifuge-chain/blob/0b6924df3d14d6c18a15f760746259baca6b47c2/pallets/foreign-investments/src/impls/mod.rs#L1076-L1083

Comment (mustermeiszer) [reply pallets/foreign-investments/src/impls/mod.rs:218]: Thanks for the details!

**Review by mustermeiszer — APPROVED** [review]
Thanks for the patience! looks good to go.
